core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -157,16 +159,31 @@ class ReplicaManager(val config: KafkaConfig,
   def recordIsrChange(topicAndPartition: TopicAndPartition) {
     isrChangeSet synchronized {
       isrChangeSet += topicAndPartition
+      lastIsrChangeMs.set(System.currentTimeMillis())
     }
   }
-
+  /**
+   * This function periodically runs to see if ISR is need to be propagated. It propagates ISR when:
+   * 1. There is ISR change not propagated yet.
+   * 2. There is no ISR Change in the last five seconds, or it has been more than 60 seconds since the last ISR propagation.
+   * This allows an occasional ISR change to be propagated within a few seconds, and avoids overwhelming controller and
+   * other brokers when large amount of ISR change occurs.
+   */
   def maybePropagateIsrChanges() {
+    val now = System.currentTimeMillis()
     isrChangeSet synchronized {
-      if (isrChangeSet.nonEmpty) {
+      if (isrChangeSet.nonEmpty &&
+        (lastIsrChangeMs.get() + 5000 < now || lastIsrPropagationMs.get() + ReplicaManager.IsrChangePropagationInterval < now)) {
         ReplicationUtils.propagateIsrChanges(zkUtils, isrChangeSet)
         isrChangeSet.clear()
+        lastIsrPropagationMs.set(now)
       }
     }
+    // If there is no ISR change to propagate, schedule the next check after three seconds.
+    // If there is pending ISR change propagation, wake up 5 seconds after the last ISR change to check if ISR change
+    // propagation is needed. In any case the next check will be after at least one second.
+    val nextCheckDelayMs = if (isrChangeSet.isEmpty) 5000 else math.max(lastIsrChangeMs.get() + 5000 - now, 1000)
```

Comment:
Would it be simpler to always schedule the next job every 5 secs? In the worse case, we may delay the propagation by 10 secs. If we care about the 5 sec delay, we can just schedule the job every 2.5 secs.